File: Assignment 5/dec2base.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
/*Student Name: Andre Wakil
ID: 260777355*/

void dec2base(int input, int base, char* str){
    //String to represent numbers in bases higher than 10 (b>10)
    char baseConversion[]="0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    //remainder and index variables declared
    int rem,i=0;
    //loop that creates string of remainders in backwards order
    while(input!=0){
        rem=input%base;
        input/=base;
        str[i]=baseConversion[rem];
        i++;
    }

}
void revStr(char* str, int length){
    //index variable declared outside for-loop because of C syntax
    int i;
    //char pointer is initialized to point to a character array (string) of length given as an argument
    char *result=(char*)malloc(sizeof(char)*(length));

    //loop 1: reverse str and store it in result
    for(i=0;i<length;i++){
        result[i]=str[length-i-1];

    }
    //loop2: copy newly-reversed char array
    for(i=0;i<length;i++){
        str[i]=result[i];
    }
    //set the character after the last index to the terminating null character so that when the string terminates
    //properly when printed in main function
    *(str+length)='\0';
}
int main(int argc, char* argv[]){
    //declaration of two integers, number and base where the sscanf function will copy the arguments from the command line
    int number, base;

    //if only one argument is passed, base 2 conversion is assumed
    if(argc==2) {
        base = 2;
        sscanf(argv[1], "%d", &number);
        //No negative values allowed
        if(number<0){
            printf("Negative number inputed.\nProgram ended");
            return 0;
        }
    }
        //if two arguments are passed (number and base), then these values are scanned by sscanf and stored in the variables number and base respectively
    else if(argc==3){
        sscanf(argv[1],"%d",&number);
        sscanf(argv[2],"%d",&base);

        if(number<0){
            printf("Negative number inputted.\nProgram ended");
            return 0;
        }
        //conversion must be done from decimal to a base in range of 2-36
        if(base>36||base<2){
            printf("Base is out of bounds. Conversion can only be done from base 2 to base 36.\nProgram ended");
            return 0;
        }
    }
        //otherwise, exit program
    else{
        printf("Wrong number of arguments\nProgram ended.");
        return 0;
    }
    //the length of the converted string is the integer rounded of (log-base(number)+1)
    //I used change of bases equation
    int length=(int)(log(number)/log(base))+1;
    char result[length];

    //dec2base called onto result (pass-by-reference)
    dec2base(number,base,result);
    //string is reversed
    revStr(result,length);
    //final result printed
    printf("The Base-%d form of %d is: %s \n",base,number, result);

}

File: Assignment 1/dec2base.java
/*Student Name: Andre Wakil
Student ID: 260777355*/
import java.util.*;//Java library imports for user input (Scanner object)

public class dec2base {

	public static void main(String[] args) {
		
		//Scanner object instantiation for user input
		Scanner userInput=new Scanner(System.in);
		System.out.println("Java base conversion demo:");
				
		//while loop that asks for a number and base to be computed until a negative number is inputed
		while(true) {
			
		System.out.print("Enter a number to be converted: ");
		//number to converted store in a variable of type int by method of the Scanner class (int nextInt())
		int num=userInput.nextInt();
		
		
		//if a negative number is inputed, the program ends and the while loop is terminated
		if(num<0) {
			System.out.println("Program ended");
			break;
		}
		//otherwise, program asks for base to convert to
		else {
			System.out.print("Enter a base to convert to: ");
			int base=userInput.nextInt();
			//Print the line that reverses the string returned by dec2B (which is in backwards order)
			System.out.println(num+" is represented in Base-"+base+" as "+reverseString(dec2B(num,base)));
		}
		}
		
	}
	//private and static method of a class that takes in two arguments, 
	//a decimal number and  a base to convert to, and returns a String of the converted string
	private static String dec2B(int number, int base) {
		//String to represent numbers in bases higher than 10 (b>10)
		String higherBases="0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
		String convertedString="";//String to add
		
		//loop that creates string of remainders in backwards order
		while(number!=0) {
			//if-else statements for when remainder is bigger than 9, 
			//then it will assign the corresponding conventional letter (A=10, B=11, C=12...)
			if(number%base>9) {

				convertedString+=higherBases.charAt(number%base);
			}
			else {
				convertedString+=(number%base);
			}
			number=number/base;
			
		}
		
		return convertedString;
	}
	//method to reverse a string by traversing it from last index(length()-1) to first index (0)
	//its purpose is to revere the backwards order string from String dec2B() 
	private static String reverseString(String result) {
		String revString="";
		
		for(int i=result.length()-1;i>=0;i--) {
			revString+=result.charAt(i);
		}
		
		return revString;
	}
}
